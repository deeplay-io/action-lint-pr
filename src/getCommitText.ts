/**
 * Regex ожидает текст вида:
 *
 * <Описание PRа для ревью>
 * ...
 * ...
 * ...
 * ***
 * <Часть, которая попадает в коммит.
 * Включает в себя описание коммита,
 * примечания и ссылку на задачу в трекере>
 */
const bodyRegex = /^.*\*{3}(.*)$/s
const commentsPattern = /(<!--.*?-->)|(<!--[\S\s]+?-->)|(<!--[\S\s]*?$)/g

export function getCommitText(prBody: string | null, prTitle: string): string {
  if (prBody == null) {
    return prTitle
  }

  const groups = prBody.match(bodyRegex)

  if (groups == null || groups[1] == null) {
    return prTitle
  }

  const resultBody = groups[1]
    // remove markdown comments
    .replace(commentsPattern, '')
    // remove leading line breaks if present
    .replace(/^(\r\n|\r|\n)+/, '')
    // remove trailing line breaks if present
    .replace(/(\r\n|\r|\n)+$/, '')

  return `${prTitle}${resultBody.length > 0 ? `\n${resultBody}` : ''}`
}
