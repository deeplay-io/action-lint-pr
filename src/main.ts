import * as core from '@actions/core'
import * as github from '@actions/github'
import '@commitlint/config-conventional'
import lint from '@commitlint/lint'
import load from '@commitlint/load'
import {LintOptions, ParserOptions, QualifiedConfig} from '@commitlint/types'
import {getCommitText} from './getCommitText'
import path from 'path'
import {existsSync} from 'fs'

const githubToken = process.env.GITHUB_TOKEN

async function run(): Promise<void> {
  if (!githubToken) {
    throw new Error('Missing github token. Check your env for GITHUB_TOKEN')
  }

  try {
    const client = github.getOctokit(githubToken)

    const contextPullRequest = github.context.payload.pull_request
    if (!contextPullRequest) {
      throw new Error(
        "This action can only be invoked in `pull_request_target` or `pull_request` events. Otherwise the pull request can't be inferred."
      )
    }

    const owner = contextPullRequest.base.user.login
    const repo = contextPullRequest.base.repo.name

    // The pull request info on the context isn't up to date. When
    // the user updates the title and re-runs the workflow, it would
    // be outdated. Therefore fetch the pull request via the REST API
    // to ensure we use the current title.
    const {data: pullRequest} = await client.rest.pulls.get({
      owner,
      repo,
      pull_number: contextPullRequest.number
    })

    const configPath = path.resolve(
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      process.env.GITHUB_WORKSPACE!,
      core.getInput('configFile')
    )
    const config = existsSync(configPath)
      ? await load({}, {file: configPath})
      : await load({extends: ['@commitlint/config-conventional']})
    await validatePrTitle(pullRequest.title, config)

    const description = getCommitText(pullRequest.body, pullRequest.title)
    core.debug(description)
    core.setOutput('commitText', description)
  } catch (error) {
    core.setFailed(error.message)
  }
}

async function validatePrTitle(
  title: string,
  config: QualifiedConfig
): Promise<void> {
  // TODO: get commitlint config from input
  // Currently blocked by @commitlint/load issue on loading configuration
  // Similar issue â€“ https://github.com/conventional-changelog/commitlint/issues/613
  const opts = getOptsFromConfig(config)
  const result = await lint(title, config.rules, opts)

  if (!result.valid) {
    throw new Error(
      `Invalid PR title: ${result.errors.map(err => `\n- ${err.message}`)}`
    )
  }
}

function getOptsFromConfig(config: QualifiedConfig): LintOptions {
  return {
    parserOpts:
      config.parserPreset != null && config.parserPreset.parserOpts != null
        ? (config.parserPreset.parserOpts as ParserOptions)
        : {},
    plugins: config.plugins != null ? config.plugins : {},
    ignores: config.ignores != null ? config.ignores : [],
    defaultIgnores: config.defaultIgnores != null ? config.defaultIgnores : true
  }
}

run()
