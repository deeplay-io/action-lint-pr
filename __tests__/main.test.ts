import {getCommitText} from '../src/getCommitText'
import {expect, test} from '@jest/globals'

const title = 'feat: test test test'
const body1 = `Test-test-test
test test
- [] do smth
- [x] test
<!--Коммент-->
***
KEY-42
<!--Коммент-->
## Примечания
<!--Коммент-->
Тест-тест-тест`
const body2 = ''
const body3 = `
- [] test
- [] bla-bla

## Примечания

`
const body4 = `
bla-bla

KEY-42



## Примечания
      Тест-тест-тест

***
<!--Коммент-->
<!--Коммент-->
`
const body5 = '\n'
const body6 = '\r'
const body7 = '\r\n'
const body8 = `
Тест

***





KEY-42
<!--Коммент-->
`

test('getCommitText matches snapshots', async () => {
  expect(getCommitText(body1, title)).toMatchSnapshot()
  expect(getCommitText(body2, title)).toMatchSnapshot()
  expect(getCommitText(body3, title)).toMatchSnapshot()
  expect(getCommitText(body4, title)).toMatchSnapshot()
  expect(getCommitText(body5, title)).toMatchSnapshot()
  expect(getCommitText(body6, title)).toMatchSnapshot()
  expect(getCommitText(body7, title)).toMatchSnapshot()
  expect(getCommitText(body8, title)).toMatchSnapshot()
})
