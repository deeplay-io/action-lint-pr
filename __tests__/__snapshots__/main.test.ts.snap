// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCommitText matches snapshots 1`] = `
"feat: test test test

KEY-42

## Примечания

Тест-тест-тест"
`;

exports[`getCommitText matches snapshots 2`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 3`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 4`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 5`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 6`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 7`] = `"feat: test test test"`;

exports[`getCommitText matches snapshots 8`] = `
"feat: test test test

KEY-42"
`;
